soot-infoflow-collections/src/soot/jimple/infoflow/collections/codeOptimization/StringResourcesResolver.java

```diff
@@ -0,0 +1,133 @@
+import soot.jimple.infoflow.InfoflowManager;
+import soot.jimple.infoflow.android.InfoflowAndroidConfiguration;
+import soot.jimple.infoflow.android.resources.ARSCFileParser;
+import soot.jimple.infoflow.codeOptimization.ICodeOptimizer;
+import soot.jimple.infoflow.entryPointCreators.SimulatedCodeElementTag;
+import soot.jimple.infoflow.solver.cfg.IInfoflowCFG;
+import soot.jimple.infoflow.sourcesSinks.manager.ISourceSinkManager;
+import soot.jimple.infoflow.taintWrappers.ITaintPropagationWrapper;
+import soot.jimple.toolkits.scalar.ConstantPropagatorAndFolder;
+import soot.util.queue.QueueReader;
+
+/**
+ * Resolves Android String resources to constants in the code.
+ *
+ * @author Tim Lange
+ */
+public class StringResourcesResolver implements ICodeOptimizer {
+    private static final String CONTEXT_CLASS = "android.content.Context";
+    private static final String GET_STRING_SUBSIG = "java.lang.String getString(int)";
+
+    private final Logger logger = LoggerFactory.getLogger(getClass());
+
+    private String fileName = null;
+    private int runtime = -1;
+    private int replaced = -1;
+
+    public int getRuntime() {
+        return runtime;
+    }
+
+    public int getReplacedStatementCount() {
+        return replaced;
+    }
+
+    @Override
+    public void initialize(InfoflowConfiguration config) {
+        if (config instanceof InfoflowAndroidConfiguration)
+            this.fileName = ((InfoflowAndroidConfiguration) config).getAnalysisFileConfig().getTargetAPKFile();
+    }
+
+    @Override
+    public void run(InfoflowManager manager, Collection<SootMethod> excluded, ISourceSinkManager sourcesSinks,
+                    ITaintPropagationWrapper taintWrapper) {
+        // This optimization enables more constant propagation. It doesn't make sense to perform it without
+        // constant propagation enabled.
+        if (manager.getConfig().getCodeEliminationMode() == InfoflowConfiguration.CodeEliminationMode.NoCodeElimination)
+            return;
+
+        // Only android apps contain resources
+        if (fileName == null)
+            return;
+
+        long beforeOptimization = System.nanoTime();
+
+        ARSCFileParser parser;
+        try {
+            parser = ARSCFileParser.getInstance(new File(fileName));
+        } catch (IOException e) {
+            logger.error("Could not parse the ARSC file! Aborting string resource resolving...", e);
+            return;
+        }
+        assert parser != null;
+
+        SootClass contextClass = Scene.v().getSootClassUnsafe(CONTEXT_CLASS);
+        if (contextClass == null) {
+            logger.error("Could not load class " + CONTEXT_CLASS + ". Aborting string resource resolving...");
+            return;
+        }
+
+        // First pass: Collect all statements to be replaced
+        ReplacementCandidates rcs = new ReplacementCandidates();
+        for (QueueReader<MethodOrMethodContext> rdr = Scene.v().getReachableMethods().listener(); rdr.hasNext();) {
```

Comment:
Wie effizient ist es, dieses Resolving vorab für alle Statements zu machen statt nur dort, wo auf die Konstante zugegriffen wird? Wäre ein lazy resolving hier nicht effizienter?

Reply:
Das hier passiert absichtlich vor der Constant Propagation, damit die resolved Konstanten direkt in den Argumenten landen. Ansonsten hätte man nicht viel gewonnen, da man trotzdem noch während der Taint Analyse ne weitere Analyse bräuchte, die einem die Konstanten zu allen Statements trägt, wo man gerne den Wert wüsste.